tuplex/core/src/ee/local/LocalBackend.cc

```diff
@@ -727,6 +742,86 @@ namespace tuplex {
         return pip_object;
     }
 
+    std::vector<std::tuple<size_t, PyObject*>> inputExceptionsToPythonObjects(std::vector<Partition *> partitions, Schema schema) {
+        using namespace tuplex;
+
+        std::vector<std::tuple<size_t, PyObject*>> pyObjects;
+        for (auto &partition : partitions) {
+            auto numRows = partition->getNumRows();
+            const uint8_t* ptr = partition->lock();
+
+            for (int i = 0; i < numRows; ++i) {
+                int64_t rowNum = *((int64_t*)ptr);
+                ptr += sizeof(int64_t);
+                int64_t ecCode = *((int64_t*)ptr);
+                ptr += 2 * sizeof(int64_t);
+                int64_t objSize = *((int64_t*)ptr);
+                ptr += sizeof(int64_t);
+
+                PyObject* pyObj = nullptr;
+                python::lockGIL();
+                if (ecCode == ecToI64(ExceptionCode::PYTHON_PARALLELIZE)) {
+                    pyObj = python::deserializePickledObject(python::getMainModule(), (char *) ptr, objSize);
+                } else {
+                    pyObj = python::rowToPython(Row::fromMemory(schema, ptr, objSize), true);
+                }
+                python::unlockGIL();
+
+                ptr += objSize;
+                pyObjects.emplace_back(rowNum, pyObj);
+            }
+
+            partition->unlock();
+            partition->invalidate();
+        }
+
+        return pyObjects;
+    }
+
+    void setExceptionInfo(const std::vector<Partition*> &normalOutput, const std::vector<Partition*> &exceptions, std::unordered_map<std::string, std::tuple<size_t, size_t, size_t>> &partitionToExceptionsMap) {
+        if (exceptions.empty()) {
+            for (const auto &p : normalOutput) {
+                partitionToExceptionsMap[uuidToString(p->uuid())] = std::make_tuple(0, 0, 0);
```

Comment:
Is uuid safe? Maybe use pointer instead?

Reply:
or make sure uuid is unique?

Reply:
prob. uuid better because then struct can be shipped across cluster e.g.

Reply:
Bit confused... do you think I should be using something other than the stringified UUID of a partition? I think I tried mapping directly from the uniqueid_t but compiler complained it wasn't hashable.

Reply:
oh I see, yeah I think that's fine. Else you can define std::hash for uuid (somewhere in base.h or so it's done for tuples).